### DISCUSSION TITLE
How to extract runoff results based on pre-defined time intervals?

### DISCUSSION THREAD

**Discussion by Freestyleyang** [pyswmm/pyswmm#510] [Q&A] [answered]
Hi There, 

Thank you for your efforts in developing the amazing package!

I am trying to extract runoff data based on pre-defined time intervals, e.g., 2020-05-10, 14:00, 2020-05-10, 14:05, ......, 2020-05-12, 15:00. 

I have tried to use `sim.step_advance(300)`, but noticed that the time stamp did not match what I need. Could you indicate how I can address this issue? Thank you.

**Comment by bemcdonnell** [comment, marked as answer]
@Freestyleyang,

When PySWMM takes on the responsibility to step a SWMM simulation forward, it calls a swmm entry-point called `swmm_step` which steps the simulation forward by one routing step.  Remember, swmm has multiple types of "time steps".  The routing step is particular to the hydraulic network and is different from the reporting time step.  Reporting time step is the expected time step you are wanting to see inside the binary output timeseries file and results are interpolated between two route steps.  I attached an example model and script to plot the time steps. See Figure 

```
REPORT_STEP          00:00:30
ROUTING_STEP         0:00:30
MINIMUM_STEP         0.05
```

There are several different types of flow routing methods that a user could choose.  The one that is most interesting for this discussion is the Dynamic Wave flow routing method in combination with variable time stepping (or routing time step).  Variable routing steps allows the "next" time step to be reduced as the model hydraulic network flow variability increases.  The benefit to variable time stepping is that during periods where the network flows are generally more stable, the time step can lengthen again and the model will run faster.  

As PySWMM interacts with the running simulation, there is generally operating system overhead for instances where PySWMM requests data out of / or pushes data to the running simulation.  Therefore, in most cases, it is not critical to interact with the running simulation at a high frequency.  By introducing the `sim.step_advance()` feature, we can reduce the simulation time by reducing our interaction frequency.  So rather than moving data to and from the model ever 2 routing step seconds, it might make more sense to do it ever 300 seconds.  This introduces a major performance gain.  

The most important statement is that pyswmm-v1 does not have discrete control over the step advance in SWMM.  This is a feature we are brining to pyswmm-v2 (hopefully the next couple weeks)!

[Archive.zip](https://github.com/pyswmm/pyswmm/files/13925173/Archive.zip)

![test](https://github.com/pyswmm/pyswmm/assets/6105134/1b4930a3-6ed9-4bbd-a6ff-646e8ad4352f)
